visualization/app/codeCharta/ui/copyToClipboardButton/util/getFilenamesWithHighestMetrics.ts

```diff
@@ -0,0 +1,41 @@
+import { hierarchy } from "d3-hierarchy"
+import { CodeMapNode, NodeType } from "../../../codeCharta.model"
+
+export type FileToValue = { name: string; value: number }
+
+const MAX_ENTRIES = 10
+
+export function getFilenamesWithHighestMetrics(node: Pick<CodeMapNode, "children" | "type" | "attributes" | "name">) {
+	const fileToValueByAttributes = new Map<string, FileToValue[]>()
+	for (const { data } of hierarchy(node)) {
+		if (data.type === NodeType.FILE && data.attributes) {
+			for (const [key, value] of Object.entries(data.attributes)) {
+				updateAttributeMap(key, value, data.name, fileToValueByAttributes)
+			}
+		}
+	}
+	return fileToValueByAttributes
+}
+
+export function updateAttributeMap(key: string, value: number, fileName: string, map: Map<string, FileToValue[]>) {
```

Comment:
## More perfomant alternativ for insertion algorithm

### Abstract

In `getFilenamesWithHighestMetric.ts` , we implemented the function `updateAttributeMap` which adds items to a `Map<string,FileToValue[]>` . This file shows how the perfomance can be improved.

### Alternative Algorithm

Instead of using the slightly costly

```
...
insertSorted(previousValues, newPair)
map.set(key, previousValues.slice(0, MAX_ENTRIES))
...
```

we could go for the following algorithm (pseudo-code):

```
IF ( previousValues.length < MAX_ENTRIES ) :
        insertSorted( previousValues, newPair )
ELSE
        insertInFullDescendingList( previousValues, newPair )
```

whereby `insertInFullDescendingList` could be implemented as

```
function insertInFullDescendingList<T extends { value: number }>(newItem: T, descendingList: T[]): void {
	let index = descendingList.length - 1
	while (index >= 0) {
		const oldItem = descendingList[index]

		if (oldItem.value < newItem.value) {
			const afterFirstLoop = index < descendingList.length - 1
			if (afterFirstLoop) {
				descendingList[index + 1] = oldItem
			}

			const atLastLoop = index === 0
			if (atLastLoop) {
				descendingList[index] = newItem
			}
		} else {
			const newValueTooSmall = index === descendingList.length - 1
			if (!newValueTooSmall) {
				descendingList[index + 1] = newItem
			}

			break
		}
		index--
	}
}
```

The perfomance would improve because we loop ONCE through the list and we IMMEDIATELY stop the iteration after `newItem` has found it's place. Also, we do not add `newItem` at all and we immediately stop the iteration, if the lowest value in the list is already bigger than `newItem.value` .